contracts/cw-governance/src/contract.rs

```diff
@@ -0,0 +1,359 @@
+        .collect();
+    if gov_module_msgs.is_empty() {
+        return Err(ContractError::NoGovernanceModule {});
+    }
+
+    GOVERNANCE_MODULE_COUNT.save(deps.storage, &(gov_module_msgs.len() as u64))?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute("sender", info.sender)
+        .add_submessage(vote_module_msg)
+        .add_submessages(gov_module_msgs))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::ExecuteProposalHook { .. } => {
+            if !GOVERNANCE_MODULES.has(deps.storage, info.sender.clone()) {
+                return Err(ContractError::Unauthorized {});
+            }
+        }
+        _ => {
+            if info.sender != env.contract.address {
+                return Err(ContractError::Unauthorized {});
+            }
+        }
+    }
+
+    let response = match msg {
+        ExecuteMsg::ExecuteProposalHook { msgs } => execute_proposal_hook(msgs),
+        ExecuteMsg::UpdateConfig { config } => execute_update_config(deps, config),
+        ExecuteMsg::UpdateVotingModule { module } => execute_update_voting_module(env, module),
+        ExecuteMsg::UpdateGovernanceModules { to_add, to_remove } => {
+            execute_update_governance_modules(deps, env, to_add, to_remove)
+        }
+        ExecuteMsg::SetItem { key, addr } => execute_set_item(deps, key, addr),
+        ExecuteMsg::RemoveItem { key } => execute_remove_item(deps, key),
+    }?;
+
+    Ok(response.add_attribute("sender", info.sender))
+}
+
+pub fn execute_proposal_hook(msgs: Vec<CosmosMsg<Empty>>) -> Result<Response, ContractError> {
+    Ok(Response::default()
+        .add_attribute("action", "execute_proposal_hook")
+        .add_messages(msgs))
+}
+
+pub fn execute_update_config(deps: DepsMut, config: Config) -> Result<Response, ContractError> {
+    CONFIG.save(deps.storage, &config)?;
+    // We incur some gas costs by having the config's fields in the
+    // response. This has the benefit that it makes it reasonably
+    // simple to ask "when did this field in the config change" by
+    // running something like `junod query txs --events
+    // 'wasm._contract_address=governance&wasm.name=name'`.
+    Ok(Response::default()
+        .add_attribute("action", "execute_update_config")
+        .add_attribute("name", config.name)
+        .add_attribute("description", config.description)
+        .add_attribute(
+            "image_url",
+            config.image_url.unwrap_or_else(|| "None".to_string()),
+        ))
+}
+
+pub fn execute_update_voting_module(
+    env: Env,
+    module: ModuleInstantiateInfo,
+) -> Result<Response, ContractError> {
+    let wasm = module.into_wasm_msg(env.contract.address);
+    let submessage = SubMsg::reply_on_success(wasm, VOTE_MODULE_UPDATE_REPLY_ID);
```

Comment:
DAOs should be able to select different voting systems for different proposals 🤔